packages/devtools/src/composite.ts

```diff
@@ -63,10 +70,68 @@ function assertModelsHaveCommits(
   }
 }
 
-function assertSupportedModelController(model: Model): void {
-  if (!model.metadata.controller.startsWith('did:key:')) {
+function assertSupportedWriteModelController(model: Model, ceramic: CeramicApi): void {
+  const unsupported = `Unsupported model controller ${model.metadata.controller}`
+  if (model.metadata.controller.startsWith('did:pkh:')){
+    if (ceramic.context.did == null) {
+      throw new Error(
+        `${unsupported}, did missing from ceramic context`
+      )
+    }
+    if (!ceramic.context.did.hasCapability()) {
+      throw new Error(
+        `${unsupported}, only did:pkh with CACAO is supported`
+      )
+    }
+    const cacao: Cacao = ceramic.context.did.capability(); 
+    if (cacao.p.exp == null) {
+      throw new Error(
+        `${unsupported}, only did:pkh CACAO without expiry is supported`
+      )
+    }
+    const hasModelResource = cacao.p.resources?.includes(`ceramic://*?model=${MODEL_STREAM_ID}`)
+    if (!hasModelResource) {
+      throw new Error(
+        `${unsupported}, only cacao with resource ${MODEL_STREAM_ID} is supported`
+      )
+    }
+  } else if (!model.metadata.controller.startsWith('did:key:')) {
     throw new Error(
-      `Unsupported model controller ${model.metadata.controller}, only did:key is supported`
+      `${unsupported}, only did:key and did:pkh are supported`
+    )
+  }
+}
+
+async function assertSupportedReadModelController(model: Model, signedCommitContainer: SignedCommitContainer): Promise<void> {
+  const unsupported = `Unsupported model controller ${model.metadata.controller}`
+  if (model.metadata.controller.startsWith('did:pkh:') && signedCommitContainer.cacaoBlock != null){
+    const cacao = await Cacao.fromBlockBytes(signedCommitContainer.cacaoBlock)
+    if (cacao == null) {
+      throw new Error(
+        `${unsupported}, only did:pkh with CACAO is supported`
+      )
+    }
+    assertValidCacao(cacao, model.metadata.controller)
+  } else if (!model.metadata.controller.startsWith('did:key:')) {
+    throw new Error(
+      `${unsupported}, only did:key is supported`
+    )
+  }
+}
+
+function assertValidCacao(cacao: Cacao, controller: string): void {
+  if (cacao.p.iss != controller) {
+    throw new Error(`Cacao issuer ${cacao.p.iss} does not match controller ${controller}`)
+  }
+  if (cacao.p.exp == null) {
```

Comment:
```suggestion
  if (cacao.p.exp !== null) {
```
This one also looks incorrect?